doc/qos/new-headroom-calculation.md

```diff
@@ -0,0 +1,168 @@
+```
+    key                     = SWITCH_PARAMETERS|<vendor name>   ; Vendor name should be in captical letters. For Mellanox, "MELLANOX"
+    cell_size               = 1*4DIGIT      ; Mandatory. The cell size of the switch chip.
+    ipg                     = 1*2DIGIT      ; Optional. Inter-packet gap.
+    mtu                     = 1*4DIGIT      ; Mandatory. Max transmit unit of RDMA packet, in unit of kBytes.
+    pipeline_latency        = 1*6DIGIT      ; Mandatory. Pipeline latency, in unit of kBytes.
+    mac_phy_delay           = 1*6DIGIT      ; Mandatory. Max/phy delay, in unit of Bytes.
+    peer_response_time      = 1*6DIGIT      ; Mandatory. The maximum of peer switch response time, in unit of kBytes.
+    other_delay             = 1*6DIGIT      ; Optional. Other delays not covered, in unit of kBytes. An example is gearbox delay.
+    small_packet_percentage = 1*3DIGIT      ; Mandatory. The percentage of small packets against all packets.
+```
+##### Initialization
+Every vendor should provide the SWITCH_PARAMETERS on a per-SKU basis. It should be stored in `/usr/shared/sonic/device/<platfrom>/<SKU>/buffers.json.j2`
+The below is an example for Mellanox switches
+***Example***
+```
+    "SWITCH_PARAMETERS": {
+        "MELLANOX": {
+            "cell_size": "144",
+            "mtu": "1.5",
+            "pipeline_latency": "18",
+            "mac_phy_delay": "0.8",
+            "peer_response_time": "3.8",
+            "small_packet_percentage": "100"
+        }
+    }
+```
+#### BUFFER_PROFILE
+Table BUFFER_PROFILE contains the profiles of headroom parameters and the proportion of free shared buffers can be utilized by a port, PG tuple (ingress) or a port, queue tuple (egress). 
+In order to support headroom override, the following filed needs to be added to the table:
+```
+    type        = "static" / "dynamic"      ; Optional. Whether the profile is dynamically calculated or user configured.
+                                            ; Default value is "static"
+```
+The `static` profile is configured by CLI while the `dynamic` profile is synamically calculated when system starts or a port's cable length or speed is updated.
+#### BUFFER_PG
+Table BUFFER_PG contains the maps from the `port, priority group` tuple to the `buffer profile` object. 
+Currently, the schema is as the following:
+```
+    key         = BUFFER_PG|<port>|<priority groups>
+    profile     = ref_hash_key_reference    ; Mandatory. Points to an entry in BUFFER_PROFILE table 
+                                            ; according to the speed and cable length of the port
+``` 
+In order to support headroom override, the following field needs to be added to the table:
+```
+    type        = "static" / "dynamic"      ; Mandatory. dynamic means a dynamically calculated buffer profile is adopted
+                                            ; while static means a static profile is adopted for headroom override. 
+```

```

Comment:
since there is DB schema change, do you need to consider DB migration?

Reply:
If we're going to downgrade from a version that supports those newly introduced fields to a version that doesn't, we need to.
Otherwise, we don't because the entries without those newly introduced fields will be treated as "dynamic" one.